### PR TITLE
Persist (remaining) remote responses locally

### PR DESCRIPTION
**Pull request by zognin**
Changes
- Persist websocket responses locally
- Persist shared document locally
- Disable import from code when offline (P.S. `isConnected` doesn't seem to be working properly :<)

### CODE REVIEW

File: Annotato/Annotato/Persistence/PersistenceService/OfflinePersistenceService/OfflinePersistenceService+AnnotationsPersistence.swift [hunk 1]
```diff
@@ -17,6 +17,10 @@ extension OfflinePersistenceService: AnnotationsPersistence {
         return await localPersistence.annotations.deleteAnnotation(annotation: annotation)
     }
 
+    func createOrUpdateAnnotation(annotation: Annotation) -> Annotation? {
```
Comment (sivayogasubramanian) [inline Annotato/Annotato/Persistence/PersistenceService/OfflinePersistenceService/OfflinePersistenceService+AnnotationsPersistence.swift]: In this layer, we set the timestamps for the other methods but we don't do that for `createOrUpdateX` functions. Is that okay?

File: Annotato/Annotato/Persistence/PersistenceService/OfflinePersistenceService/OfflinePersistenceService+DocumentsPersistence.swift [hunk 2]
```diff
@@ -29,6 +29,10 @@ extension OfflinePersistenceService: DocumentsPersistence {
         return await localPersistence.documents.deleteDocument(document: document)
     }
 
+    func createOrUpdateDocument(document: Document) -> Document? {
```
Comment (sivayogasubramanian) [inline Annotato/Annotato/Persistence/PersistenceService/OfflinePersistenceService/OfflinePersistenceService+DocumentsPersistence.swift]: Same thing about not setting the timestamps.

File: Annotato/Annotato/Persistence/PersistenceService/OnlinePersistenceService/OnlinePersistenceService+AnnotationsPersistence.swift [hunk 3]
```diff
@@ -14,6 +14,10 @@ extension OnlinePersistenceService: AnnotationsPersistence {
         await remotePersistence.annotations.deleteAnnotation(annotation: annotation)
     }
 
+    func createOrUpdateAnnotation(annotation: Annotation) -> Annotation? {
```
Comment (sivayogasubramanian) [inline Annotato/Annotato/Persistence/PersistenceService/OnlinePersistenceService/OnlinePersistenceService+AnnotationsPersistence.swift:47]: Why are the methods on lines 17 and 21 calling local persistence when it is in the `OnlinePersistenceService+AnnotationsPersistence.swift` file?

File: Annotato/Annotato/Persistence/PersistenceService/OnlinePersistenceService/OnlinePersistenceService+AnnotationsPersistence.swift [hunk 4]
```diff
@@ -14,6 +14,10 @@ extension OnlinePersistenceService: AnnotationsPersistence {
...
+        localPersistence.annotations.createOrUpdateAnnotation(annotation: annotation)
+    }
+
```
Comment (sivayogasubramanian) [inline Annotato/Annotato/Persistence/PersistenceService/OnlinePersistenceService/OnlinePersistenceService+AnnotationsPersistence.swift:51]: Unrelated to this PR but the methods on lines 5, 9 and 13 should be calling remote persistence first and then getting the response and saving it to the local persistence same as `OnlinePersistenceService+DocumentsPersistence`.

File: Annotato/Annotato/Persistence/PersistenceService/OnlinePersistenceService/OnlinePersistenceService+AnnotationsPersistence.swift [hunk 5]
```diff
@@ -14,6 +14,10 @@ extension OnlinePersistenceService: AnnotationsPersistence {
         await remotePersistence.annotations.deleteAnnotation(annotation: annotation)
     }
 
+    func createOrUpdateAnnotation(annotation: Annotation) -> Annotation? {
+        localPersistence.annotations.createOrUpdateAnnotation(annotation: annotation)
```
Comment (sivayogasubramanian) [inline Annotato/Annotato/Persistence/PersistenceService/OnlinePersistenceService/OnlinePersistenceService+AnnotationsPersistence.swift]:
```suggestion
        remotePersistence.annotations.createOrUpdateAnnotation(annotation: annotation)
```

File: Annotato/Annotato/Persistence/PersistenceService/OnlinePersistenceService/OnlinePersistenceService+DocumentSharesPersistence.swift [hunk 6]
```diff
@@ -2,12 +2,14 @@ import AnnotatoSharedLibrary
...
-        guard let createdDocumentShareRemote = await remotePersistence
+    func createDocumentShare(documentShare: DocumentShare) async -> Document? {
+        guard let sharedDocument = await remotePersistence
             .documentShares
             .createDocumentShare(documentShare: documentShare) else {
             return nil
         }
-        return createdDocumentShareRemote
+
+        _ = await localPersistence.documents.createDocument(document: sharedDocument)
```
Comment (sivayogasubramanian) [inline Annotato/Annotato/Persistence/PersistenceService/OnlinePersistenceService/OnlinePersistenceService+DocumentSharesPersistence.swift]: Let's return the result from here for consistency with `OnlinePersistenceService+DocumentsPersistence`. Unless there is a good reason to do it this way.

File: Annotato/Annotato/Persistence/PersistenceService/OnlinePersistenceService/OnlinePersistenceService+DocumentsPersistence.swift [hunk 7]
```diff
@@ -52,6 +52,10 @@ extension OnlinePersistenceService: DocumentsPersistence {
         return deletedDocumentLocal
     }
 
+    func createOrUpdateDocument(document: Document) -> Document? {
+        localPersistence.documents.createOrUpdateDocument(document: document)
```
Comment (sivayogasubramanian) [inline Annotato/Annotato/Persistence/PersistenceService/OnlinePersistenceService/OnlinePersistenceService+DocumentsPersistence.swift]: Same thing as previous comment for lines 55 and 59.

```suggestion
        remotePersistence.documents.createOrUpdateDocument(document: document)
```

File: Annotato/Annotato/Utils/NetworkMonitor.swift [hunk 8]
```diff
@@ -1,10 +1,11 @@
 import Foundation
 import Network
+import Combine
 
 class NetworkMonitor {
```
Comment (sivayogasubramanian) [inline Annotato/Annotato/Utils/NetworkMonitor.swift]: Should this conform to `ObservableObject`?

File: Annotato/Annotato/Views/Document/List/DocumentListImportMenu.swift [hunk 9]
```diff
@@ -70,4 +75,12 @@ class DocumentListImportMenu: UIStackView {
...
         self.isHidden = true
     }
+
+    private func setUpSubscribers() {
+        NetworkMonitor.shared.$isConnected.sink(receiveValue: { [weak self] isConnected in
+            DispatchQueue.main.async {
+                self?.fromCodeButton.isEnabled = isConnected
+            }
+        }).store(in: &cancellables)
+    }
```
Comment (sivayogasubramanian) [inline Annotato/Annotato/Views/Document/List/DocumentListImportMenu.swift:85]: This seems to work for me quite well actually. Why did you say that the isConnected thing is not working properly? How to reproduce?

File: Annotato/Annotato/Persistence/PersistenceService/OnlinePersistenceService/OnlinePersistenceService+AnnotationsPersistence.swift (hunk 3 above)
Comment (zognin) [reply Annotato/Annotato/Persistence/PersistenceService/OnlinePersistenceService/OnlinePersistenceService+AnnotationsPersistence.swift:47]: online persistence service is supposed to call both local and remote, but remote one isn't implemented (and not necessary for our app). so it only calls the local one.

File: Annotato/Annotato/Views/Document/List/DocumentListImportMenu.swift (hunk 9 above)
Comment (zognin) [reply Annotato/Annotato/Views/Document/List/DocumentListImportMenu.swift:85]: I enter the app with wifi connected, then I disconnect the wifi, up to here is correct. But when I turn on the wifi again, `isConnected` doesn't becomes true (it logs as false).

File: Annotato/Annotato/Persistence/PersistenceService/OfflinePersistenceService/OfflinePersistenceService+AnnotationsPersistence.swift [hunk 10]
```diff
@@ -17,7 +17,23 @@ extension OfflinePersistenceService: AnnotationsPersistence {
         return await localPersistence.annotations.deleteAnnotation(annotation: annotation)
     }
 
-    func createOrUpdateAnnotations(annotations: [Annotation]) -> [Annotation]? {
-        localPersistence.annotations.createOrUpdateAnnotations(annotations: annotations)
+    func createOrUpdateAnnotation(annotation: Annotation) async -> Annotation? {
+        if LocalAnnotationEntityDataAccess.read(annotationId: annotation.id,
```
Comment (sivayogasubramanian) [inline Annotato/Annotato/Persistence/PersistenceService/OfflinePersistenceService/OfflinePersistenceService+AnnotationsPersistence.swift:21]: I think the `OfflinePersistenceService` should not be made aware of the DataAccess layer. Instead, it should just go through `localPersistence.annotations.getAnnotation` or something. What do you think?

File: Annotato/Annotato/Persistence/PersistenceService/OfflinePersistenceService/OfflinePersistenceService+DocumentsPersistence.swift [hunk 11]
```diff
@@ -29,7 +29,23 @@ extension OfflinePersistenceService: DocumentsPersistence {
...
-    func createOrUpdateDocuments(documents: [Document]) -> [Document]? {
-        localPersistence.documents.createOrUpdateDocuments(documents: documents)
+    func createOrUpdateDocument(document: Document) async -> Document? {
+        if LocalDocumentEntityDataAccess.read(documentId: document.id,
+                                              withDeleted: true) != nil {
+            return await updateDocument(document: document)
+        } else {
+            return await createDocument(document: document)
+        }
+    }
```
Comment (sivayogasubramanian) [inline Annotato/Annotato/Persistence/PersistenceService/OfflinePersistenceService/OfflinePersistenceService+DocumentsPersistence.swift:39]: Same as the above comment.

```suggestion
    func createOrUpdateDocument(document: Document) async -> Document? {
        if await localPersistence.documents.getDocument(documentId: document.id) != nil {
            return await updateDocument(document: document)
        } else {
            return await createDocument(document: document)
        }
    }
```

**Review by sivayogasubramanian — COMMENTED** [review]
LGTM! We will check later during the meeting.

**Pull request merged**